src/core/control/jobs/RenderJob.cpp

```diff
@@ -77,14 +78,32 @@ void RenderJob::run() {
 
         renderToBuffer(newBuffer.get());
 
-        std::lock_guard lock(this->view->drawingMutex);
-        std::swap(this->view->crBuffer, newBuffer);
+        {
+            std::lock_guard lock(this->view->drawingMutex);
+            std::swap(this->view->crBuffer, newBuffer);
+        }
+        repaintPage();
     } else {
-        for (Rectangle<double> const& rect: rerenderRects) { rerenderRectangle(rect); }
+        for (Rectangle<double> const& rect: rerenderRects) {
+            rerenderRectangle(rect);
+            repaintPageArea(rect.x, rect.y, rect.x + rect.width, rect.y + rect.height);
+        }
     }
+}
+
+static void repaintWidgetArea(GtkWidget* widget, int x1, int y1, int x2, int y2) {
+    Util::execInUiThread([=]() { gtk_xournal_repaint_area(widget, x1, y1, x2, y2); });
+}
+
+void RenderJob::repaintPage() const {
+    repaintWidgetArea(view->xournal->getWidget(), 0, 0, view->getWidth(), view->getHeight());
```

Comment:
Maybe I'm missing something, but won't this repaint the area around the top left corner of the canvas widget instead of the top left corner of the page? I think this should be:
```suggestion
    repaintPageArea(view->xournal->getWidget(), 0, 0, view->getWidth(), view->getHeight());
```